ipaclient/plugins/vault.py

```diff
@@ -568,14 +580,119 @@ def forward(self, *args, **options):
         return response
 
 
+class _TransportCertInvalid(Exception):
+    def __init__(self, exc_info):
+        self.exc_info = exc_info
+
+
+_transport_cert_cache = {}
+
+
+class ModVaultData(Local):
+    def _do_internal(self, mechanism, session_key, transport_cert_der,
+                     *args, **options):
+        nss_transport_cert = nss.Certificate(transport_cert_der)
+
+        # wrap session key with transport certificate
+        # pylint: disable=no-member
+        public_key = nss_transport_cert.subject_public_key_info.public_key
+        # pylint: enable=no-member
+        wrapped_session_key = nss.pub_wrap_sym_key(mechanism,
+                                                   public_key,
+                                                   session_key)
+
+        options['session_key'] = wrapped_session_key.data
+
+        name = self.name + '_internal'
+        try:
+            return self.api.Command[name](*args, **options)
+        except errors.NotFound:
+            raise
+        except (errors.InternalError,
+                errors.ExecutionError,
+                errors.GenericError):
+            raise _TransportCertInvalid(sys.exc_info())
```

Comment:
```sys.exc_info()``` is problematic. It returns a stack frame object, which usually creates a reference cycle that keeps all local objects alive. If any object has ```__del__``` method, then Python 2 is unable to break the reference cycle. It's rather bad for long running processes because it is a kind of memory leak.

Here are two reasonable well written blog postings that I just found
* http://engineering.hearsaysocial.com/2013/06/16/circular-references-in-python/
* https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/

Proposal:

```
class _TransportCertInvalid(Exception):
    def __init__(self, exception):
        self.inner = exception
```

and

```
        except (errors.InternalError,
                     errors.ExecutionError,
                     errors.GenericError) as e:
            root_logger.debug("some message", exc_info=True)
            raise _TransportCertInvalid(e)
```

later:

```
six.reraise(type(e.inner), e.inner, None)
```